### PR TITLE
Add check for duplicated doc aliases

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @estebank

### CODE REVIEW

**Comment by joshtriplett** [comment]
Does this check only for duplicate aliases on one item, or could it also check for duplicates throughout a crate? I would love to see a warning lint for the latter.

**Comment by GuillaumeGomez** [comment]
The latter is perfectly fine though. You can wrap a same C function differently a few times so I definitely don't think that we should warn about it.

File: compiler/rustc_passes/src/check_attr.rs [hunk 1]
```diff
@@ -582,17 +590,38 @@ impl CheckAttrVisitor<'tcx> {
         if &*item_name.as_str() == doc_alias {
             return err_fn(meta.span(), "is the same as the item's name");
         }
+        let span = meta.span();
+        if aliases.try_insert(doc_alias.to_owned(), span).is_err() {
+            self.tcx.struct_span_lint_hir(UNUSED_ATTRIBUTES, hir_id, span, |lint| {
+                lint.build("doc alias is duplicated")
+                    .span_label(*aliases.get(doc_alias).unwrap(), "first defined here")
```
Comment (Mark-Simulacrum) [inline compiler/rustc_passes/src/check_attr.rs:597]: nit: you don't need the second lookup here, right? You can do `if let Err(entry) = ...` above and then `*entry.get()` here?

Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:597]: Absolutely! Great catch!

Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:597]: Ah actually no, I did read it well the first time: it returns the non-inserted value, which is useless in here.

Comment (Mark-Simulacrum) [reply compiler/rustc_passes/src/check_attr.rs:597]: *entry.entry.get() should still work: the returned Err contains both the value passed in and the hashmap entry you would have overwritten with a regular insert.

Comment (GuillaumeGomez) [reply compiler/rustc_passes/src/check_attr.rs:597]: Damn, didn't pay attention to that. This API is even better than what I originally thought!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by joshtriplett** [comment]
@GuillaumeGomez Libs, at least, has a policy against doing that, so it'd be helpful to have a lint we could turn on to enforce that.

**Comment by GuillaumeGomez** [comment]
No, even libs don't want it: [example](https://doc.rust-lang.org/nightly/std/collections/hash_map/struct.HashMap.html?search=!).

**Comment by joshtriplett** [comment]
Fair enough, it wouldn't be universal. It might be useful as a warning to be opted out of, but in any case that's not a matter for this PR.

**Comment by GuillaumeGomez** [comment]
I think a rustc-specific lint might be the way to go. But yes, out of scope here. ;)

**Pull request merged**